### PR TITLE
add pnp loss

### PR DESCRIPTION
**Pull request by interestingzhuo**
add pnp loss proposed in aaai2022
"https://github.com/interestingzhuo/PNPloss"

### CODE REVIEW

**Comment by KevinMusgrave** [comment]
Thanks @interestingzhuo !

For the test, can you compare the result with your original implementation? Just paste your original loss function (https://github.com/interestingzhuo/PNPloss/blob/main/criteria/PNP.py) into the test file, and then in the test, compare the output of the pytorch-metric-learning version with the output of your original version.

**Comment by interestingzhuo** [comment]
> Thanks @interestingzhuo !
> 
> For the test, can you compare the result with your original implementation? Just paste your original loss function (https://github.com/interestingzhuo/PNPloss/blob/main/criteria/PNP.py) into the test file, and then in the test, compare the output of the pytorch-metric-learning version with the output of your original version.

Done

File: src/pytorch_metric_learning/losses/pnp_loss.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        I_pos[a1_idx, p_idx] = 1
+        I_neg[a2_idx, n_idx] = 1
+        N_pos = torch.sum(I_pos, dim=1)
+        safe_N = N_pos > 0
+        if torch.sum(safe_N) == 0:
+            return self.zero_losses()
+        sim_all = self.compute_aff(embeddings)
+
+        
+        mask = I_neg.unsqueeze(dim=1).repeat(1, N, 1).cuda()
```
Comment (KevinMusgrave) [inline src/pytorch_metric_learning/losses/pnp_loss.py:39]: Remove .cuda()

File: src/pytorch_metric_learning/losses/pnp_loss.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+        elif self.variant == "PNP-I_b":
+            b = self.b
+            sim_all_rk = 1 / b**2 * (b * sim_all_rk - torch.log(1 + b * sim_all_rk))
+        elif self.variant == "PNP-O":
+            pass
+        else:
+            raise Exception("variantation <{}> not available!".format(self.variant))
+        
+        loss = (sim_all_rk * I_pos.cuda()) / N_pos.reshape(-1, 1)
```
Comment (KevinMusgrave) [inline src/pytorch_metric_learning/losses/pnp_loss.py:65]: Remove .cuda()

Comment (interestingzhuo) [reply src/pytorch_metric_learning/losses/pnp_loss.py]: Done

File: src/pytorch_metric_learning/losses/pnp_loss.py (hunk 1 above)
Comment (interestingzhuo) [reply src/pytorch_metric_learning/losses/pnp_loss.py]: Done

File: tests/losses/test_pnp_loss.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            return 0
+        self.mask = I_neg.unsqueeze(dim=1).repeat(1, N, 1)
+        
+        sim_all = self.compute_aff(batch)
+        sim_all = sim_all 
+        sim_all_repeat = sim_all.unsqueeze(dim=1).repeat(1, N, 1)
+        # compute the difference matrix
+        sim_diff = sim_all_repeat - sim_all_repeat.permute(0, 2, 1)
+        # pass through the sigmoid and ignores the relevance score of the query to itself
+        sim_sg = self.sigmoid(sim_diff, temp=self.anneal) * self.mask.cuda()
```
Comment (KevinMusgrave) [inline tests/losses/test_pnp_loss.py:44]: Also remove this cuda call

File: tests/losses/test_pnp_loss.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        elif self.variant == 'PNP-I_b':
+            b = self.b
+            sim_all_rk = 1/b**2 * (b*sim_all_rk-torch.log(1+b*sim_all_rk))
+        elif self.variant == 'PNP-O':
+            pass
+        else:
+                raise Exception('variantation <{}> not available!'.format(self.variant))
+        
+        
+        loss = (sim_all_rk * I_pos.cuda()) / N_pos.reshape(-1, 1)
```
Comment (KevinMusgrave) [inline tests/losses/test_pnp_loss.py:65]: and remove this cuda call

File: tests/losses/test_pnp_loss.py (hunk 3 above)
Comment (interestingzhuo) [reply tests/losses/test_pnp_loss.py]: done

File: tests/losses/test_pnp_loss.py (hunk 4 above)
Comment (interestingzhuo) [reply tests/losses/test_pnp_loss.py]: done

**Comment by KevinMusgrave** [comment]
@interestingzhuo  I noticed the `OriginalImplementationPNP` that you provided in `test_pnp_loss` is different from the implementation in your repo (https://github.com/interestingzhuo/PNPloss). Is there a reason for that? I would prefer to compare to the actual original implementation.

**Comment by interestingzhuo** [comment]
> @interestingzhuo I noticed the `OriginalImplementationPNP` that you provided in `test_pnp_loss` is different from the implementation in your repo (https://github.com/interestingzhuo/PNPloss). Is there a reason for that? I would prefer to compare to the actual original implementation.

To fit the functional interface, I modified the original implementation. The implementation is tested  in the SOP dataset.

**Comment by KevinMusgrave** [comment]
So the implementation that uses functions like `lmu.get_all_pairs_indices(labels)` is the implementation that you carefully tested on SOP?

**Comment by KevinMusgrave** [comment]
I guess my point is that `OriginalImplementationPNP` looks almost identical to `PNPLoss`. I just want to make sure we're comparing `PNPLoss` to the true original implementation.

**Comment by KevinMusgrave** [comment]
Yeah I'd prefer if you replace `OriginalImplementationPNP` with the implementation in https://github.com/interestingzhuo/PNPloss, and then in `test_pnp_loss`, just reshape `embeddings` when you pass it into `original_loss_func`

**Comment by KevinMusgrave** [comment]
@interestingzhuo Does my approach of reshaping the embeddings make sense for testing the original implementation? I don't mind making the change myself if you're too busy.

**Comment by interestingzhuo** [comment]
> @interestingzhuo Does my approach of reshaping the embeddings make sense for testing the original implementation? I don't mind making the change myself if you're too busy.

Yeah,  I'm busy these weeks, you can easily reshape the input and test it.

**Comment by KevinMusgrave** [comment]
@interestingzhuo I formatted the code and made some small fixes. The test passes now. If it looks ok to you then I'll merge.

**Comment by interestingzhuo** [comment]
> @interestingzhuo I formatted the code and made some small fixes. The test passes now. If it looks ok to you then I'll merge.
I have checked it, you can directly merge it. Thank you!
